sphinx/_cli/__init__.py

```diff
@@ -0,0 +1,298 @@
+"""Base 'sphinx' command.
+
+Subcommands are loaded lazily from the ``_COMMANDS`` table for performance.
+
+All subcommand modules must define three attributes:
+
+- ``parser_description``, a description of the subcommand. The first paragraph
+  is taken as the short description for the command.
+- ``set_up_parser``, a callable taking and returning an ``ArgumentParser``. This
+  function is responsible for adding options and arguments to the subcommand's
+  parser.
+- ``run``, a callable taking parsed arguments and returning an exit code. This
+  function is responsible for running the main body of the subcommand and
+  returning the exit status.
+
+The entire ``sphinx._cli`` namespace is private, only the command line interface
+has backwards-compatability guarantees.
+"""
+
+from __future__ import annotations
+
+import argparse
+import locale
+import sys
+from typing import TYPE_CHECKING
+
+from sphinx._cli.util.colour import (
+    bold,
+    disable_colour,
+    enable_colour,
+    terminal_supports_colour,
+    underline,
+)
+from sphinx.locale import __, init_console
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterable, Iterator, Sequence
+    from typing import NoReturn
+
+    _PARSER_SETUP = Callable[[argparse.ArgumentParser], argparse.ArgumentParser]
+    _RUNNER = Callable[[argparse.Namespace], int]
+
+    if sys.version_info[:2] > (3, 8):
+        from typing import Protocol
+
+        class _SubcommandModule(Protocol):
+            parser_description: str
+            set_up_parser: _PARSER_SETUP  # takes and returns argument parser
+            run: _RUNNER  # takes parsed args, returns exit code
+    else:
+        from types import ModuleType as _SubcommandModule
+
+
+# Map of command name to import path.
+_COMMANDS: dict[str, str] = {
```

Comment:
Nothing seems to populate this. I assume that's because this isn't complete? Mind marking the PR as a draft until things are closer to shippable?

Reply:
See comment below re series of PRs -- I intend to merge this one first, and then have follow-ups to implement each individual command.